File: front-end/src/services/StudentService.ts
import { Student } from '../core/entities/student.entity';

export interface StudentService {
  getAllStudents(): Promise<Student[]>;
  getStudentById(id: number): Promise<Student>;
  getStudentWithDetails(id: number): Promise<Student>;
  createStudent(studentData: any): Promise<Student>;
  updateStudent(id: number, studentData: any): Promise<Student>;
  deleteStudent(id: number): Promise<void>;
  findBySchoolId(schoolId: number): Promise<Student[]>;
  findByParentId(parentId: number): Promise<Student[]>;
  findByStudentId(studentId: string): Promise<Student>;
  findByQrCode(qrCode: string): Promise<Student>;
  searchStudents(query: string): Promise<Student[]>;
  assignToSchool(studentId: number, schoolId: number): Promise<void>;
  assignToParent(studentId: number, parentId: number): Promise<void>;
  removeFromSchool(studentId: number): Promise<void>;
  removeFromParent(studentId: number): Promise<void>;
  generateQrCode(studentId: number): Promise<void>;
  updateStudentDetails(studentId: number, firstName: string, lastName: string, grade: string): Promise<void>;
  findStudentsWithAttendanceStats(schoolId: number, start: string, end: string): Promise<Student[]>;
  countAbsences(studentId: number, start: string, end: string): Promise<number>;
  calculateAttendancePercentage(studentId: number, start: string, end: string): Promise<number>;
  findStudentsByRoute(routeId: number): Promise<Student[]>;
} 

File: back-end/src/main/java/com/example/backend/service/StudentService.java
package com.example.backend.service;

import com.example.backend.entities.Student;
import com.example.backend.service.base.BaseService;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

public interface StudentService extends BaseService<Student> {
    List<Student> findBySchoolId(Long schoolId);
    List<Student> findByParentId(Long parentId);
    Optional<Student> findByQrCode(String qrCode);
    Optional<Student> findByIdWithDetails(Long id);
    List<Student> searchStudents(String query);
    void assignToSchool(Long studentId, Long schoolId);
    void assignToParent(Long studentId, Long parentId);
    void removeFromSchool(Long studentId);
    void removeFromParent(Long studentId);
    void generateQrCode(Long studentId);
    void updateStudentDetails(Long studentId, String firstName, String lastName, String grade);
    List<Student> findStudentsWithAttendanceStats(Long schoolId, LocalDateTime start, LocalDateTime end);
    long countAbsences(Long studentId, LocalDateTime start, LocalDateTime end);
    double calculateAttendancePercentage(Long studentId, LocalDateTime start, LocalDateTime end);
    List<Student> findStudentsByRoute(Long routeId);
} 